Google.Api.Generator/ProtoUtils/ResourceDetails.cs

```diff
@@ -174,54 +204,92 @@ string GetShortName(ResourceDescriptor resDesc)
+                    var parentPatterns = childDef.Patterns.Select(x => ParentPattern(x.PatternString)).ToImmutableList();
+                    var parentPatternsSet = parentPatterns.ToImmutableHashSet();
+                    var parentDefs = parentPatterns
+                        .SelectMany(x => resourcesByPattern.TryGetValue(x, out var defs) ? defs : Enumerable.Empty<Definition>())
+                        .Where(def => def.Patterns.All(x => parentPatternsSet.Contains(x.PatternString)))
+                        .Distinct(Definition.TypeComparer.Instance)
+                        .OrderBy(def => def.Patterns.Count)
+                        .ThenBy(def => def.Patterns.Select(x => parentPatterns.IndexOf(x.PatternString)).Where(x => x >= 0).Min())
+                        .ThenBy(def => def.UnifiedResourceTypeName)
+                        .ToImmutableList();
+                    // "*" pattern is not required to be matched; all others are required.
+                    if (parentPatternsSet.Except(parentDefs.SelectMany(x => x.Patterns).Select(x => x.PatternString).Append("*")).Any())
+                    {
+                        throw new InvalidOperationException($"Not all patterns can be matched to a parent resource for field {msgDesc.Name}.{fieldDesc.Name}");
+                    }
+                    foreach (var parentDef in parentDefs)
+                    {
+                        yield return new Field(fieldDesc, parentDef);
+                    }
+                    if (parentDefs.Any(x => x.HasWildcard) || parentDefs.Count > 1)
+                    {
+                        yield return new Field(fieldDesc, Definition.WildcardResource, parentDefs.Count > 1 ? parentDefs : null, parentPatternsSet.Contains("*"));
+                    }
                 }
-                var parentPatterns = childDef.Patterns.Select(x => ParentPattern(x.PatternString)).ToImmutableHashSet();
-                if (resourcesByPatterns.TryGetValue(parentPatterns, out var parentDef))
+                else
                 {
-                    // Return existing resource; no auto-generated required.
-                    return new Field(fieldDesc, parentDef);
+                    throw new InvalidOperationException("type or child_type must be set.");
                 }
-                // It is invalid to ask for a parent that is not already defined.
-                // This may change in the future, to allow auto-generating parent resource-names, but this is not currently allowed.
-                throw new InvalidOperationException(
-                    $"Cannot refer to child-type '{resourceRef.ChildType}' in field {msgDesc.Name}.{fieldDesc.Name} because the child patterns are not already defined in a resource.");
             }
-            throw new InvalidOperationException("type or child_type must be set.");
 
             string ParentPattern(string pattern)
             {
+                if (pattern == "*")
```

Comment:
Ooh, is this something I forgot to do in a previous PR? Sorry.

Reply:
I don't think so, it was the simplest way to get multi-parent-including-wildcard-patterns working ;)